src/daft-local-execution/src/sinks/grouped_aggregate.rs

```diff
@@ -0,0 +1,301 @@
+use std::{cmp::min, sync::Arc};
+
+use common_error::DaftResult;
+use common_runtime::RuntimeRef;
+use daft_core::prelude::SchemaRef;
+use daft_dsl::{col, Expr, ExprRef};
+use daft_micropartition::MicroPartition;
+use daft_physical_plan::extract_agg_expr;
+use tracing::{info_span, instrument, Instrument};
+
+use super::blocking_sink::{
+    BlockingSink, BlockingSinkFinalizeResult, BlockingSinkSinkResult, BlockingSinkState,
+    BlockingSinkStatus,
+};
+use crate::NUM_CPUS;
+
+struct SinglePartitionAggregateState {
+    partially_aggregated: Vec<MicroPartition>,
+    unaggregated: Vec<MicroPartition>,
+    unaggregated_size: usize,
+}
+
+enum GroupedAggregateState {
+    Accumulating {
+        inner_states: Vec<Option<SinglePartitionAggregateState>>,
+        agg_exprs: Option<Vec<ExprRef>>,
+        group_by: Vec<ExprRef>,
+    },
+    Done,
+}
+
+impl GroupedAggregateState {
+    // This is the threshold for when we should aggregate the unaggregated partitions
+    const PARTIAL_AGG_THRESHOLD: usize = 10_000;
+    // This is the maximum number of partitions we can have
+    const MAX_NUM_PARTITIONS: usize = 16;
```

Comment:
why do we cap this rather than always take num cores? 

Not a suggestion, just curious

Reply:
Because it would be bad for low cardinality aggs if there's too many partitions. But I modified the algorithm to cater for low cardinality aggs, so this is removed.